src/gammalasso.jl

```diff
@@ -0,0 +1,406 @@
+      γ=convert(Vector{T},repmat([γ],p))
+    else
+      @assert length(γ)==p "length(γ) != number of parameters ($p)"
+    end
+
+    # initialize penalty factors to 1 (no rescaling to sum to the number of coefficients)
+    ω = ones(T,p)
+    # ω = rescale(ω,p)
+
+    # Lasso initialization
+    α = convert(T, α)
+    λminratio = convert(T, λminratio)
+    coefitr = randomize ? RandomCoefficientIterator() : (1:0)
+    cd = naivealgorithm ? NaiveCoordinateDescent{T,intercept,typeof(X),typeof(coefitr)}(X, α, maxncoef, 1e-7, coefitr, ω) :
+                          CovarianceCoordinateDescent{T,intercept,typeof(X),typeof(coefitr)}(X, α, maxncoef, 1e-7, coefitr, ω)
+
+    # GLM response initialization
+    autoλ = λ == nothing
+    wts .*= convert(T, 1/sum(wts))
+    off = convert(Vector{T}, offset)
+
+    if isa(d, Normal) && isa(l, IdentityLink)
+        # Special no-IRLS case
+        mu = isempty(offset) ? y : y + off
+        nullb0 = intercept ? mean(mu, weights(wts)) : zero(T)
+        nulldev = 0.0
+        @simd for i = 1:length(mu)
+            @inbounds nulldev += abs2(mu[i] - nullb0)*wts[i]
+        end
+
+        if autoλ
+            # Find max λ
+            if intercept
+                muscratch = Array(T, length(mu))
+                @simd for i = 1:length(mu)
+                    @inbounds muscratch[i] = (mu[i] - nullb0)*wts[i]
+                end
+            else
+                muscratch = mu.*wts
+            end
+            Xy = X'muscratch
+            λ = computeλ(Xy, λminratio, α, nλ, ω)
+        else
+            λ = convert(Vector{T}, λ)
+        end
+
+        # First y is just a placeholder here
+        model = LinearModel(LmResp{typeof(y)}(mu, off, wts, y), cd)
+    else
+        # Fit to find null deviance
+        # Maybe we should reuse this GlmResp object?
+        nullmodel = fit(GeneralizedLinearModel, ones(T, n, ifelse(intercept, 1, 0)), y, d, l;
+                        wts=wts, offset=offset, convTol=irls_tol, dofit=dofit)
+        nulldev = deviance(nullmodel)
+
+        if autoλ
+            # Find max λ
+            Xy = X'*broadcast!(*, nullmodel.rr.wrkresid, nullmodel.rr.wrkresid, nullmodel.rr.wrkwts)
+            λ = computeλ(Xy, λminratio, α, nλ, ω)
+            nullb0 = intercept ? coef(nullmodel)[1] : zero(T)
+        else
+            λ = convert(Vector{T}, λ)
+            nullb0 = zero(T)
+        end
+
+        eta = GLM.initialeta!(d, l, similar(y), y, wts, off)
+        rr = GlmResp{typeof(y),typeof(d),typeof(l)}(y, d, l, eta, similar(eta), offset, wts)
+        model = GeneralizedLinearModel(rr, cd, false)
+    end
+
+    # Fit path
+    path = GammaLassoPath{typeof(model),T}(model, nulldev, nullb0, λ, autoλ, γ, Xnorm)
+    if dofit
+        fit!(path; irls_tol=irls_tol, fitargs...)
+    else
+        path.λ = zeros(T, 0)
+        path.pct_dev = zeros(T, 0)
+        path.coefs = spzeros(T, p, 0)
+        path.b0 = zeros(T, 0)
+        path.niter = 0
+    end
+    path
+end
```

Comment:
Is there a way to cut down on the amount of repetition of code from LassoPath here?